quixstreams/dataframe/joins/join_interval.py

```diff
@@ -0,0 +1,72 @@
+import typing
+from datetime import timedelta
+from typing import Any, Callable, Optional, Union
+
+from quixstreams.dataframe.utils import ensure_milliseconds
+
+from .base import Join, JoinHow, OnOverlap
+
+if typing.TYPE_CHECKING:
+    from quixstreams.dataframe.dataframe import StreamingDataFrame
+
+__all__ = ("IntervalJoin",)
+
+
+class IntervalJoin(Join):
+    def __init__(
+        self,
+        how: JoinHow,
+        on_merge: Union[OnOverlap, Callable[[Any, Any], Any]],
+        grace_ms: Union[int, timedelta],
+        store_name: Optional[str] = None,
+        backward_ms: Union[int, timedelta] = 0,
+        forward_ms: Union[int, timedelta] = 0,
+    ):
+        super().__init__(how, on_merge, grace_ms, store_name)
+        self._backward_ms = ensure_milliseconds(backward_ms)
+        self._forward_ms = ensure_milliseconds(forward_ms)
+
+        if self._backward_ms > self._grace_ms:
+            raise ValueError(
+                "The backward_ms must not be greater than the grace_ms "
+                "to avoid losing data."
+            )
+
+    def _prepare_join(
+        self,
+        left: "StreamingDataFrame",
+        right: "StreamingDataFrame",
+    ) -> "StreamingDataFrame":
+        self._register_store(left, keep_duplicates=True)
+        self._register_store(right, keep_duplicates=True)
+
+        tx = self._get_transaction
+        is_inner_join = self._how == "inner"
+        merger = self._merger
+        backward_ms = self._backward_ms
+        forward_ms = self._forward_ms
+
+        def left_func(value, key, timestamp, headers):
+            tx(left).set_for_timestamp(timestamp=timestamp, value=value, prefix=key)
+
+            if right_values := tx(right).get_interval(
+                start=timestamp - backward_ms,
+                end=timestamp + 1,  # +1 because end is exclusive
+                prefix=key,
+            ):
+                return [merger(value, right_value) for right_value in right_values]
```

Comment:
`merger` may mutate the value, so I think we should pass copies of `value` here and in the `right_func` respectively. Still scratching my head on this, WDYT @daniil-quix ?
```suggestion
                return [merger(value.copy(), right_value) for right_value in right_values]
```